lib/pure/complex.nim

```diff
@@ -9,78 +9,150 @@
 
 
 
+
+
 ## This module implements complex numbers.
-{.push checks:off, line_dir:off, stack_trace:off, debugger:off.}
-# the user does not want to trace a part
-# of the standard library!
+## Complex numbers are currently implemented as generic on a 64-bit or 32-bit float.
 
+{.push checks: off, line_dir: off, stack_trace: off, debugger: off.}
+# the user does not want to trace a part of the standard library!
 
-import
-  math
 
-const
-  EPS = 1.0e-7 ## Epsilon used for float comparisons.
+import math
 
 type
-  Complex* = tuple[re, im: float]
-    ## a complex number, consisting of a real and an imaginary part
+  Complex*[T: SomeFloat] = tuple[re, im: T]
```

Comment:
Not backward compatible but I would prefer an object instead of a tuple here.

RFC? :D

Reply:
Actually yes. This should be an object, I agree even though it is a breaking change. 
Only an object would allow a later optimization with SIMD. SIMD requires self alignment, and alignment can only be specified on object types.

Reply:
Switching to objects means we couldn't do `z: Complex128 = (0.0, 0.0)` anymore, and we should have `z = Complex128(re: 0.0, im: 0.0)`. This leaves us with:

```nim
var
  a = Complex128(re: 1.0, im: 2.0)
  b = a + Complex128(re: 3.0, im: 4.0)
```

instead of current:
```nim
var
  a: Complex128 = (1.0, 2.0)
  b = a + (3.0, 4.0)
```

Are we sure we want that?

Reply:
there should be still this (untested) syntax:

```nim
var
  a = 1.0 + 2.0 * im
  b = a + 3.0 + 4.0 * im
```

I don't like the breaking change, but the old api was not very type safe. I disagree that any tuple of numbers is also a complex numbler. On top of that I think it is a good idea, if toComplex128 and toComplex64 is overloaded so that it also accepts a number for the imaginary part.

Reply:
Usually we create constructor procs like ``complex64(0.0, 1.0)``